src/interpreter.rs

```diff
@@ -120,3 +129,1207 @@ bitflags::bitflags! {
+                            if stack.size() < 4 {
+                                return set_error(ScriptError::InvalidStackOperation);
+                            }
+                            let vch1 = stack.top(-4)?.clone();
+                            let vch2 = stack.top(-3)?.clone();
+                            stack.push_back(vch1);
+                            stack.push_back(vch2);
+                        }
+
+                        OP_2ROT => {
+                            // (x1 x2 x3 x4 x5 x6 -- x3 x4 x5 x6 x1 x2)
+                            if stack.size() < 6 {
+                                return set_error(ScriptError::InvalidStackOperation);
+                            }
+                            let vch1 = stack.top(-6)?.clone();
+                            let vch2 = stack.top(-5)?.clone();
+                            stack.erase(stack.end() - 6, Some(stack.end() - 4));
+                            stack.push_back(vch1);
+                            stack.push_back(vch2);
+                        }
+
+                        OP_2SWAP => {
+                            // (x1 x2 x3 x4 -- x3 x4 x1 x2)
+                            if stack.size() < 4 {
+                                return set_error(ScriptError::InvalidStackOperation);
+                            }
+                            stack.swap(-4, -2)?;
+                            stack.swap(-3, -1)?;
+                        }
+
+                        OP_IFDUP => {
+                            // (x - 0 | x x)
+                            if stack.size() < 1 {
+                                return set_error(ScriptError::InvalidStackOperation);
+                            }
+                            let vch = stack.top(-1)?;
+                            if cast_to_bool(vch) {
+                                stack.push_back(vch.to_vec())
+                            }
+                        }
+
+                        OP_DEPTH => {
+                            // -- stacksize
+                            let bn = ScriptNum::try_from(stack.size()).map_err(|_| ScriptError::StackSize)?;
+                            stack.push_back(bn.getvch())
+                        }
+
+                        OP_DROP => {
+                            // (x -- )
+                            if stack.size() < 1 {
+                                return set_error(ScriptError::InvalidStackOperation);
+                            }
+                            stack.pop()?;
+                        }
+
+                        OP_DUP => {
+                            // (x -- x x)
+                            if stack.size() < 1 {
+                                return set_error(ScriptError::InvalidStackOperation);
+                            }
+
+                            let a = stack.pop()?;
+                            stack.push_back(a.clone());
+                            stack.push_back(a);
```

Comment:
Why not more like the C++?
```suggestion
                            let vch = stack.top(-1)?;
                            stack.push_back(vch.clone());
```

Reply:
This suggestion appears to have an error; it pushes once rather than twice.

Reply:
That being said, I think that using `pop` is clearer; I don't know whether `.top` modifies the stack.

Reply:
> I think that using pop is clearer [...]

Perhaps, but I asked because this PR is supposed to reflect the C++ code very closely.